src/transformers/modeling_utils.py

```diff
@@ -833,21 +852,40 @@ def _generate_no_beam_search(
                 next_token = torch.argmax(next_token_logits, dim=-1)
 
             # update generations and finished sentences
-            tokens_to_add = next_token * unfinished_sents + pad_token_id * (1 - unfinished_sents)
+            if eos_token_ids is not None:
+                # if pad_token_id is undefined, use first eos_token_id since sentence will be filled with -1 anyways
+                tokens_to_add = next_token * unfinished_sents + (pad_token_id) * (1 - unfinished_sents)
+            else:
+                tokens_to_add = next_token
+
             input_ids = torch.cat([input_ids, tokens_to_add.unsqueeze(-1)], dim=-1)
-            for eos_token_id in eos_token_ids:
-                unfinished_sents.mul_(tokens_to_add.ne(eos_token_id).long())
+
+            if eos_token_ids is not None:
+                for eos_token_id in eos_token_ids:
+                    eos_in_sents = tokens_to_add == eos_token_id
+                    # batches_len is updated only the first time eos_token is in the sentences (unfinished_sents == 1 and eos_in_sents == 1)
+                    batches_len.masked_fill_(unfinished_sents.mul(eos_in_sents.long()).bool(), cur_len + 1)
+                    # unfinished_sents is set to zero if eos in sentence
+                    unfinished_sents.mul_((~eos_in_sents).long())
+
             cur_len = cur_len + 1
 
             # stop when there is a </s> in each sentence, or if we exceed the maximul length
             if unfinished_sents.max() == 0:
                 break
 
-        # add eos_token_ids to unfinished sentences
-        if cur_len == max_length:
-            input_ids[:, -1].masked_fill_(unfinished_sents.to(dtype=torch.bool), eos_token_ids[0])
+        # if some batches have to be padded, the pad_token_id has to be defined
+        if batches_len.min().item() != batches_len.max().item():
+            assert pad_token_id is not None, "`Pad_token_id` has to be defined if `eos_token_ids`"
```

Comment:
I would move this assert much earlier, as previously mentioned.

Reply:
I changed the error message and commetn to make it a bit clearer! 

I like the assert here - it's just there to verify again that if sentences have different lengths, then some sentences have to be padded which requires `pad_token_id` to be defined.